### PR TITLE
feat(ImageMapper): When XYZ slicing, display intermediary slices

### PR DESCRIPTION
**Pull request by boucaud**
### PR and Code Checklist

- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Context

This fixes https://discourse.vtk.org/t/vtkimageslice-xyz-slicing-and-image-position/6673

And discussion at the 9/23 dev meeting.
### Changes

Change getSliceAtPosition to not round slice values.

Change the vtkOpenGLImageMapper to round slice values before accessing scalars.

- [x] Documentation and TypeScript definitions were updated to match those changes

### Results

Slice positions are no longer rounded when slicing with X/Y/Z.
### Testing

- [x] This change adds or fixes unit tests
- [x] All tests complete without errors on the following environment:
  - **vtk.js**: 
  - **OS**: Linux
  - **Browser**: chromium 93.0.4577.82

### CODE REVIEW

**Comment by boucaud** [comment]
@agirault @finetjul  WDYT ?

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 1]
```diff
@@ -46,24 +48,24 @@ function vtkImageMapper(publicAPI, model) {
...
       case SlicingMode.J:
         slice = vtkMath.clampValue(ijk[1], ex[2], ex[3]);
-        slice = Math.round(slice);
         break;
       case SlicingMode.K:
         slice = vtkMath.clampValue(ijk[2], ex[4], ex[5]);
-        slice = Math.round(slice);
         break;
       default:
         return 0;
```
Comment (finetjul) [inline Sources/Rendering/Core/ImageMapper/index.js:63]: return `null` (or `vtkErrorMacro`) instead of 0 which is a valid slice?

**Comment by finetjul** [comment]
LGTM

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 2]
```diff
@@ -46,24 +48,24 @@ function vtkImageMapper(publicAPI, model) {
...
     }
+
+    if (ijkMode === model.slicingMode) {
+      slice = Math.round(slice);
+    }
```
Comment (agirault) [inline Sources/Rendering/Core/ImageMapper/index.js:68]: Why redo the rounding here if the goal is to do continuous index?

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 3]
```diff
@@ -22,19 +22,21 @@ function vtkImageMapper(publicAPI, model) {
...
           break;
-        case SlicingMode.Y:
+        case SlicingMode.J:
           pos3 = [(bds[1] + bds[0]) / 2, pos, (bds[5] + bds[4]) / 2];
           break;
-        case SlicingMode.Z:
+        case SlicingMode.K:
           pos3 = [(bds[1] + bds[0]) / 2, (bds[3] + bds[2]) / 2, pos];
           break;
         default:
```
[the remark this answers is not in the document]
Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js:42]: Example: you have no guarantee that your K axis will be along the Z axis, which is needed to compute `pos3` properly here.

File: Sources/Rendering/Core/ImageMapper/test/testSlicingModePosition.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import vtkImageMapper from 'vtk.js/Sources/Rendering/Core/ImageMapper';
+import { SlicingMode } from 'vtk.js/Sources/Rendering/Core/ImageMapper/Constants';
+import vtkImageSlice from 'vtk.js/Sources/Rendering/Core/ImageSlice';
+import vtkRTAnalyticSource from 'vtk.js/Sources/Filters/Sources/RTAnalyticSource';
+
+test('Test slice position differences between XYZ and IJK modes', (t) => {
+  const source = vtkRTAnalyticSource.newInstance({
+    wholeExtent: [0, 10, 0, 10, 0, 10],
+  });
```
Comment (agirault) [inline Sources/Rendering/Core/ImageMapper/test/testSlicingModePosition.js:11]: Test with a spacing != 1 and an orientation that's not the identity if you want to test differences between XYZ and IJK.

File: Sources/Rendering/Core/ImageMapper/index.d.ts [hunk 5]
```diff
@@ -34,10 +34,11 @@ interface IImageMapperInitialValues {
...
-	/**
-	 *
-	 * @param {Number} pos
-	 */
+  /**
+   * Get the slice index from pos.
+   * When slicing with XYZ, returns intermediary slice positions instead of IJK indices.
+   * @param {number[] | number} [pos] World point or XYZ slice value
+   */
 	getSliceAtPosition(pos: number): number;
```
Comment (agirault) [inline Sources/Rendering/Core/ImageMapper/index.d.ts:43]: `getSliceAtPosition` is meant to convert a world position to a slice index. The current slicing mode is only used when passing a single number, so it goes from 1d to 3d based on the current axis. This method should not be used to pass coordinates in index space.

**A couple of suggestions:**
1. Improve current doc to illustrate `getSliceAtPosition` 
2. Consider renaming? (`getSliceIndexAtWorldPosition`)
3. Consider a separate method to get the continuous index so we don't break too many assumptions in apps that depend on `getSliceAtPosition` ? Similar to ITK (ex: `getContinuousSliceIndexAtWorldPosition` ... ugh). No strong feelings on this atm.

File: Sources/Rendering/Core/ImageMapper/index.js (hunk 2 above)
Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js:68]: As discussed:
- this rounding only occurred if `slicingMode` was I, J or K
- however, this method should really only be called when in X, Y or Z (it does world to index given one single XYZ component, outputting one single IJK component). yes, the name of the method is terrible in hindsight
- either way, we'd want IJK to be able to be continuous also, and just fix the mapper to round the index before accessing the scalar data, so no rounding needed here anyway.

File: Sources/Rendering/Core/ImageMapper/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js:42]: As discussed: reverting those changes. Ideally, `getSliceAtPosition` should ensure the slicing mode is X, Y or Z (won't work for IJK). A better name might also be `getContinuousSliceIndexAtWorldPosition`.

File: Sources/Rendering/Core/ImageMapper/test/testSlicingModePosition.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  mapper.setSlicingMode(SlicingMode.K);
+
+  t.equal(3, mapper.getSliceAtPosition(3));
+  t.equal(3, mapper.getSliceAtPosition(3.1));
+
+  t.equal(5, mapper.getSliceAtPosition([0, 0, 5]));
+  t.equal(5, mapper.getSliceAtPosition([0, 0, 5.1]));
+
+  t.equal(10, mapper.getSliceAtPosition(15));
+  t.equal(10, mapper.getSliceAtPosition(15.1));
```
Comment (agirault) [inline Sources/Rendering/Core/ImageMapper/test/testSlicingModePosition.js:39]: `getSliceAtPosition` should only be used with XYZ slicing mode.

**Comment by agirault** [comment]
@jadh4v could you review?

**Comment by jadh4v** [comment]
LGTM

**Comment by finetjul** [comment]
@agirault can you please remove the "changes requested" ?

**Review by agirault — APPROVED** [review]
My bad. Thanks for the heads up @finetjul, and for the changes @boucaud !

**Comment by boucaud** [comment]
@floryst I am having issues figuring out why CI is failing for this new test. The images provided as output seem identical, and the test passes when I test locally (both with and without https://github.com/Kitware/vtk-js/pull/2098).

Do you have an idea what could be causing this ?

**Comment by floryst** [comment]
Not 100% sure yet, but it stems from using `run-tests-in-docker.sh`. The output looks like this:

image

I wonder if it's an issue with the opengl container that runs the tests.

**Comment by boucaud** [comment]
> Not 100% sure yet, but it stems from using `run-tests-in-docker.sh`. The output looks like this:
> 
> image
> 
> I wonder if it's an issue with the opengl container that runs the tests.

I fixed the issue by explicitly setting the ColorTransferFunction to do what I needed instead of relying on window/level.

I'm still not sure why the previous version would produce different results in different testing environments.

**Comment by boucaud** [comment]
@finetjul I believe this can be merged now

**Comment by floryst** [comment]
Thanks, that resolved the test. I may need to look into why the original test was failing at a later point.

**Pull request merged**